src/main/java/com/commercetools/sync/producttypes/ProductTypeSync.java

```diff
@@ -0,0 +1,279 @@
+    private static final String PRODUCT_TYPE_DRAFT_IS_NULL = "Failed to process null product type draft.";
+
+    private final ProductTypeService productTypeService;
+
+    public ProductTypeSync(@Nonnull final ProductTypeSyncOptions productTypeSyncOptions) {
+        super(new ProductTypeSyncStatistics(), productTypeSyncOptions);
+        this.productTypeService = new ProductTypeServiceImpl(productTypeSyncOptions);
+    }
+
+    public ProductTypeSync(@Nonnull final ProductTypeSyncOptions productTypeSyncOptions,
+                           @Nonnull final ProductTypeService productTypeService) {
+        super(new ProductTypeSyncStatistics(), productTypeSyncOptions);
+        this.productTypeService = productTypeService;
+    }
+
+    /**
+     * Iterates through the whole {@code productTypeDrafts} list and accumulates its valid drafts to batches.
+     * Every batch is then processed by {@link ProductTypeSync#processBatch(List)}.
+     *
+     * <p><strong>Inherited doc:</strong>
+     * {@inheritDoc}
+     *
+     * @param productTypeDrafts {@link List} of {@link ProductTypeDraft}'s that would be synced into CTP project.
+     * @return {@link CompletionStage} with {@link ProductTypeSyncStatistics} holding statistics of all sync
+     *         processes performed by this sync instance.
+     */
+    @Override
+    protected CompletionStage<ProductTypeSyncStatistics> process(
+            @Nonnull final List<ProductTypeDraft> productTypeDrafts) {
+
+        final List<List<ProductTypeDraft>> batches = batchElements(productTypeDrafts, syncOptions.getBatchSize());
+
+        return syncBatches(batches, CompletableFuture.completedFuture(statistics));
+    }
+
+    @Override
+    protected CompletionStage<ProductTypeSyncStatistics> syncBatches(
+            @Nonnull final List<List<ProductTypeDraft>> batches,
+            @Nonnull final CompletionStage<ProductTypeSyncStatistics> result) {
+
+        if (batches.isEmpty()) {
+            return result;
+        }
+
+        final List<ProductTypeDraft> firstBatch = batches.remove(0);
+        return syncBatches(batches, result.thenCompose(subResult -> processBatch(firstBatch)));
+    }
+
+    /**
+     * Fetches existing {@link ProductType} objects from CTP project that correspond to passed {@code batch}.
+     * Having existing product types fetched, {@code batch} is compared and synced with fetched objects by
+     * {@link ProductTypeSync#syncBatch(List, List)} function. When fetching existing product types results in
+     * an empty optional then {@code batch} isn't processed.
+     *
+     * @param batch batch of drafts that need to be synced
+     * @return {@link CompletionStage} of {@link Void} that indicates method progress.
+     */
+    @Override
+    protected CompletionStage<ProductTypeSyncStatistics> processBatch(@Nonnull final List<ProductTypeDraft> batch) {
+        final List<ProductTypeDraft> validProductTypeDrafts = batch.stream()
+                .filter(this::validateDraft)
```

Comment:
please always indent the . chains on the same line, so it's easier to read:

````java
        final List<ProductTypeDraft> validProductTypeDrafts = batch.stream()
                                                                   .filter(this::validateDraft)
                                                                   .collect(toList());
````

Reply:
:) I change it but my understanding when we discussed about was that the second chained function and successive could be indented in the beginning on the next line, because otherwise the code is moving to the right each time we have a chaining function with body, until you finally reach the right gutter where you don't have enough space (i.e https://github.com/commercetools/commercetools-sync-java/blob/master/src/main/java/com/commercetools/sync/categories/CategorySync.java#L175 , the arrow functions don't fit in one line making the code tough to read because the chain of function is almost in the right end)

And I completely agree that the `.` should be aligned but my understanding was from the second chained function onwards :)

Reply:
959790a37dcf9d23270261d97dc20f663d44f01d

Reply:
no, then we had a misunderstanding 😁 What I meant is if it's too hard to read like the example you sent, then start the first chain from the beginning of the next line, like:


````java

return categoryService
.cacheKeysToIds()
.thenCompose(keyToIdCache -> {
  prepareDraftsForProcessing(categoryDrafts, keyToIdCache);
  categoryKeysToFetch = existingCategoryDrafts.stream().map(CategoryDraft::getKey)
  ...
````

and in this case:

````java
final List<ProductTypeDraft> validProductTypeDrafts = batch
.stream()
.filter(this::validateDraft)
.collect(toList());
````

but honestly, in that case I think it's still possible to have the chains on the same indentation while keeping the first chain on the same line (like I mentioned in the comment before):

````java
final List<ProductTypeDraft> validProductTypeDrafts = batch.stream()
                                                           .filter(this::validateDraft)
                                                           .collect(toList());
````

WDYT?